### PR TITLE
chore: add metris for memtable read path

### PR DESCRIPTION
**Pull request by v0y4g3r**
I hereby agree to the terms of the [GreptimeDB CLA](https://gist.github.com/xtang/6378857777706e568c1949c7578592cc)

## What's changed and what's your intention?

This PR adds following metrics for the read path of MergeTreeMemtable:

MERGE_TREE_DATA_BUFFER_FREEZE_STAGE_ELAPSED
- DataPartEncoder
    - drain_data_buffer_to_record_batches (buffer_to_batch)
    - ArrowWriter::write (encode)

MERGE_TREE_READ_STAGE_ELAPSED
- TreeIter
    - TreeIter::fetch_next_partition (fetch_next_partition)
    - TreeIter::next_batch
        - PartitionReader::convert_current_batch (partition_data_batch_to_batch)
        - PartitionReader::prune_batch_by_key (partition_prune_pk)
        - PartitionReader::next (partition_read_source)
            - ShardBuilderReader::next (*)
                - DataBufferRead::next (read_data_buffer)
            - ShardReader::next
                - DataPartsReader::next (read_data_parts)
                    - DataBuffer::read (data_buffer_to_batch)
                    - DataPart::read (read_data_part) : mostly spend on reading parquet bytes

## Checklist

- [ ]  I have written the necessary rustdoc comments.
- [ ]  I have added the necessary unit tests and integration tests.
- [x]  This PR does not require documentation updates.

### CODE REVIEW

File: src/mito2/src/memtable/merge_tree/tree.rs [hunk 1]
```diff
@@ -316,6 +319,15 @@ impl MergeTree {
 struct TreeIter {
     partitions: VecDeque<PartitionRef>,
     current_reader: Option<PartitionReader>,
+    fetch_partition_elapsed: Duration,
```
Comment (evenyag) [inline src/mito2/src/memtable/merge_tree/tree.rs:322]: Could we also collect how many rows and batches we returned? We might also log them in debug logs.

File: src/mito2/src/memtable/merge_tree/partition.rs [hunk 2]
```diff
@@ -224,6 +226,13 @@ pub(crate) struct PartitionStats {
     pub(crate) shared_memory_size: usize,
 }
 
+#[derive(Default)]
+struct PartitionReaderMetrics {
+    prune_pk: Duration,
```
Comment (evenyag) [inline src/mito2/src/memtable/merge_tree/partition.rs:231]: Do we need to count number of keys before and after pruning?

File: src/mito2/src/memtable/merge_tree/data.rs [hunk 3]
```diff
@@ -523,6 +540,7 @@ impl DataBufferReader {
 
     /// Advances reader to next data batch.
     pub(crate) fn next(&mut self) -> Result<()> {
+        let start = Instant::now();
         if self.offset >= self.batch.num_rows() {
             self.current_range = None;
             return Ok(());
```
[the remark this answers is not in the document]
Comment (v0y4g3r) [reply src/mito2/src/memtable/merge_tree/data.rs:545]: fixed https://github.com/GreptimeTeam/greptimedb/pull/3397/commits/8a1736bc7931ec9ffaf44962aff46713c5f0bc91#diff-5066fcdd3b887926278b1155cfbf005cbca27e9699e36292abcc55ecf80110e2R547

File: src/mito2/src/memtable/merge_tree/partition.rs (hunk 2 above)
Comment (v0y4g3r) [reply src/mito2/src/memtable/merge_tree/partition.rs:232]: > Do we need to count number of keys before and after pruning?

fixed https://github.com/GreptimeTeam/greptimedb/pull/3397/commits/8a1736bc7931ec9ffaf44962aff46713c5f0bc91#diff-5066fcdd3b887926278b1155cfbf005cbca27e9699e36292abcc55ecf80110e2R547
not sure if these are the correct place to collect pruned keys.

File: src/mito2/src/memtable/merge_tree/tree.rs (hunk 1 above)
Comment (v0y4g3r) [reply src/mito2/src/memtable/merge_tree/tree.rs]: > Could we also collect how many rows and batches we returned? We might also log them in debug logs.

fixed https://github.com/GreptimeTeam/greptimedb/pull/3397/commits/8a1736bc7931ec9ffaf44962aff46713c5f0bc91#diff-c49c8c96fe96144915d2d855220956e3284752035404a1a9b230934e2ec213deR389-R390

File: src/mito2/src/memtable/merge_tree/data.rs [hunk 4]
```diff
@@ -911,7 +953,10 @@ impl DataParts {
             nodes.push(DataNode::new(DataSource::Part(p.read()?)));
         }
         let merger = Merger::try_new(nodes)?;
-        Ok(DataPartsReader { merger })
+        Ok(DataPartsReader {
```
[the remark this answers is not in the document]
Comment (v0y4g3r) [reply src/mito2/src/memtable/merge_tree/data.rs:960]: > We might need to add the cost time of building the DataPartsReader to `elapsed` or the metrics MERGE_TREE_READ_STAGE_ELAPSED.

fixed https://github.com/GreptimeTeam/greptimedb/pull/3397/commits/8a1736bc7931ec9ffaf44962aff46713c5f0bc91#diff-5066fcdd3b887926278b1155cfbf005cbca27e9699e36292abcc55ecf80110e2R946-R948

**Review by evenyag — APPROVED** [review]
LGTM

File: src/mito2/src/memtable/merge_tree/shard_builder.rs [hunk 5]
```diff
@@ -107,10 +108,21 @@ impl ShardBuilder {
...
+        };
 
+        {
+            let _timer = MERGE_TREE_READ_STAGE_ELAPSED
+                .with_label_values(&["sort_pk"])
+                .start_timer();
+            dict_reader.pk_weights_to_sort_data(pk_weights_buffer);
+        }
+
+        let data_reader = self.data_buffer.read(Some(pk_weights_buffer))?;
```
Comment (waynexia) [inline src/mito2/src/memtable/merge_tree/shard_builder.rs:125]: Would this be costly?

**Review by waynexia — APPROVED** [review]
LGTM

File: src/mito2/src/memtable/merge_tree/shard_builder.rs (hunk 5 above)
Comment (v0y4g3r) [reply src/mito2/src/memtable/merge_tree/shard_builder.rs:125]: > Would this be costly?

Cost time of `self.data_buffer.read` will be recorded here: https://github.com/GreptimeTeam/greptimedb/blob/8a1736bc7931ec9ffaf44962aff46713c5f0bc91/src/mito2/src/memtable/merge_tree/data.rs#L259-L263

**Pull request merged**